apps/x509.c

```diff
@@ -1103,8 +1165,17 @@ static int callb(int ok, X509_STORE_CTX *ctx)
 /* self-issue; self-sign unless a forced public key (fkey) is given */
 static int sign(X509 *x, EVP_PKEY *pkey, EVP_PKEY *fkey, int days, int clrext,
                 const EVP_MD *digest, CONF *conf, const char *section,
-                int preserve_dates)
+                int preserve_dates, STACK_OF(OPENSSL_STRING) *sigopts)
 {
+#ifndef OPENSSL_NO_SM2
+    if (EVP_PKEY_id(pkey) == EVP_PKEY_SM2) {
+        /*
+         * sm2 need to set sm2_id before sign, sm2_id should be set to
+         * EVP_PKEY_CTX, so use func do_X509_sign to do sm2 sign
+         */
+        return do_X509_sign(x, pkey, digest, sigopts);
+    }
+#endif
```

Comment:
Instead if this special case, it should be enough to replace `X509_sign(x, pkey, digest)` with `do_X509_sign(x, pkey, digest, sigopts)` below